### PR TITLE
[AuditV2] Load instance analytics into h2

### PR DESCRIPTION
**Pull request by snoe**
Part of #31632

UPDATE: `CASE_INSENSITIVE_IDENTIFIERS` didn't pan out, see comment below for explanation. The approach taken here is to load the lowercased serdes data from postgres into h2 by ensuring that fields and tables match casing during load, then sets casing back to what a sync would find.

---
This is an interesting run around the problem of h2 normally treating identifiers as case sensitive. H2 v 2 introduced CASE_INSENSITIVE_IDENTIFIERS https://www.h2database.com/javadoc/org/h2/engine/DbSettings.html

Normally, when you create table foobar in H2, that table will be created with uppercase identifiers FOOBAR however, when you select from foobar the table will not be found as identifiers need to match the case in which they are stored during queries by default. This setting allows unquoted identifiers in queries to match table and column names in a case insensitive way.

What this means, is that we can serialize from a postgres db and restore in an h2 db. Table and field names will be loaded as lowercase, and the queries will be built with lowercase identifiers but still match the created uppercase tables and field names. This is good because it should allow native queries to be portable.

Unfortunately, the setting does not work for schema identifiers. So before and after loading we have to update the schema names to lower then upper case.

### CODE REVIEW

**Comment by camsaul** [comment]
Left notes in thread https://metaboat.slack.com/archives/C056QEW4KNF/p1692382714275499 here 

In principle this change is ok with me, altho the H2 dox does say it is experimental, which I don't love, as long as this is purely a query engine thing and works with existing databases (as opposed to only affecting newly created stuff). In MySQL for comparison case-sensitivity in string comparison (collation) is set at table creation time and changing it on existing tables requires DB migrations. So you double-check the behavior for existing DBs here.

I'm also questioning whether or not we really need this since option we already have ways of dealing with this stuff...

H2 already treats unquoted identifiers as upper-case, so if you want something to work between Postgres and H2 you can just leave it unquoted. e.g.

```sql
SELECT * FROM core_user;
```

will work in both Postgres and H2.

You can test this behavior out yourself:

```clj
(dev/query-jdbc-db :h2 "SELECT * FROM venues;") ; => works
(dev/query-jdbc-db :h2 "SELECT * FROM VENUES;") ; => works
```

If you want to quote the identifiers/are using Honey SQL you can just use our custom :metabase.db.setup/application-db dialect that will do the right thing for H2 vs Postgres e.g.0

```clj
(sql/format honey-sql {:quoted true, :dialect :metabase.db.setup/application-db, :quoted-snake false})
```

 then I think this makes sense, we don't have any identifiers in the application database that use mixed case anyway so this will just make our lives a bit easier.

So maybe you should check whether we can/should just leverage the existing stuff instead, or why that wouldn't work for some reason.

Let me know what you find, if this works out maybe we can use this everywhere and drop our custom Honey SQL dialect stuff entirely

File: enterprise/backend/src/metabase_enterprise/audit_db.clj [hunk 1]
```diff
@@ -75,21 +75,41 @@
     (if-let [audit-db (t2/select-one :model/Database {:where [:= :is_audit true]})]
       (do (log/info "Beginning Audit DB Sync...")
           (log/with-no-logs (sync-metadata/sync-db-metadata! audit-db))
-          (log/info "Audit DB Sync Complete."))
+          (log/info "Audit DB Sync Complete.")
+          ;; We need to move back to a schema that matches the serialized data
+          (when (= :h2 mdb.env/db-type)
+            (t2/update! :metabase_database (:id audit-db) {:engine "postgres"})
```
Comment (camsaul) [inline enterprise/backend/src/metabase_enterprise/audit_db.clj:81]:
```suggestion
            (t2/update! (t2/table-name :model/Database) (:id audit-db) {:engine "postgres"})
```

One of these days we're finally going to rename these tables to something more sensible, and we don't want to have to go update 1000 LoC in the app code

**Review by camsaul — APPROVED** [review]
I don't really understand what's going on here (the PR description seems to be for the old version of the PR before the latest changes)

but if I read this correctly, is it changing all the Databases/Tables/Fields for the Audit DB to lowercase, loading stuff into them with SerDes v2, then changing all those Databases/Tables/Fields back to uppercase? 

I don't really love this solution, it seems like the right way to fix this would have been just to handle the name transformation inside SerDes itself, but I guess I don't want the the perfect be the enemy of the good here, so I guess I'm ok with this as long as we open an issue to revisit this in the future.

**Comment by snoe** [comment]
> I don't really understand what's going on here (the PR description seems to be for the old version of the PR before the latest changes)
> 
> but if I read this correctly, is it changing all the Databases/Tables/Fields for the Audit DB to lowercase, loading stuff into them with SerDes v2, then changing all those Databases/Tables/Fields back to uppercase?
> 
> I don't really love this solution, it seems like the right way to fix this would have been just to handle the name transformation inside SerDes itself, but I guess I don't want the the perfect be the enemy of the good here, so I guess I'm ok with this as long as we open an issue to revisit this in the future.

In the original issue, there was an attempt to do the name change with the serdes incoming data and that quickly becomes  unmanageable with the different ways things can be referenced.

I looked into doing the transforms within serdes themselves. I thought the place to do the transforms would be in the various serdes defmethods in `models.field` and `model.table` and then in `model.card` that references them. But it felt like it added a lot of branching and context carrying, maybe doubling a lot of that code. I think if we wanted a destination agnostic deserializer that's one thing, but I think this approach is going to be a lot more straightforward at the cost of 6 extra writes on top of the serdes work.

**Comment by snoe** [comment]
`CASE_INSENSITIVE_IDENTIFIERS`
What this setting does is actually make quoted identifiers case insensitive. I picked up the issue thinking the problem was that `select * from venues` didn't work in H2 but, in fact the issue is that `select * from "venues"` is the query that's being generated. I don't believe we want quoted identifiers to be case insensitive.

Changing the custom dialect to drop quoting ran into broken "expected identifier" syntax errors right away. I didn't look too closely into this because more importantly syncing was broken.

In the original approach, serdes would create table & field entries with lowercase `SCHEMA.table.field`, and while querying the loaded cards would find the correct fields and running a query that would find H2's `SCHEMA.TABLE.FIELD`, sync would create duplicate entires of `SCHEMA.TABLE.FIELD`.

**Pull request merged**